### PR TITLE
stm32h7: support SDRAM via FMC peripherial

### PR DESCRIPTION
**Pull request by anpaza**
This patchset contains three commits:

- f047619 fixes out-of-tree board directory support, making it work like it was intended (according to docs). I couldn’t test my later patches without this one, as I use a non-standard board for development.
- d6213fe is a small patch that prevents NuttX from crashing if developer sets MM_REGIONS too low.
- 051e74e contains mostly changes to stm32h7/stm32_fmc.{c,h} and stm32_allocateheap.c. Also it fixes a serious typo in stm32h7x3xx_rcc.h which caused PLL2 and PLL3 not to work at all on stm32h7.

### CODE REVIEW

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 1]
```diff
@@ -39,40 +39,387 @@
...
+
 #include "stm32.h"
 
-#if defined(CONFIG_STM32H7_FMC)
+#include <debug.h>
+
+//#include "stm32_fmc.h"
+//#include "hardware/stm32_pinmap.h"
+//#include "stm32_gpio.h"
+//#include "stm32_rcc.h"
```
Comment (patacongo) [inline arch/arm/src/stm32h7/stm32_fmc.c]: Use of commented out code and any use of C++ comments is forbidden by the coding standard.

**Comment by davids5** [comment]
@anpaza - Thank you. These all have great merit. But the coupling is problematic. In looking at this it should be 3 PRs for the 3 commits.

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 2]
```diff
@@ -39,40 +39,387 @@
...
+
+#ifdef HAVE_FMC_SDRAM
+  stm32_fmc_sdram_init();
+#endif
+
+  /* Set up SDRAM refresh timings */
+
+#ifdef BOARD_FMC_SDRAM_REFR_PERIOD
+  /* ... The programmed COUNT value must not be equal to the sum of the following timings:
+     TWR+TRP+TRC+TRCD+4 memory clock cycles */
```
Comment (patacongo) [inline arch/arm/src/stm32h7/stm32_fmc.c]: */ must be on a separate line per coding standard
All comments must be followed by a blank line per coding standard.

File: mm/mm_heap/mm_initialize.c [hunk 3]
```diff
@@ -75,6 +75,13 @@ void mm_addregion(FAR struct mm_heap_s *heap, FAR void *heapstart,
   uintptr_t heapend;
 #if CONFIG_MM_REGIONS > 1
   int IDX = heap->mm_nregions;
+
+  /* Writing past CONFIG_MM_REGIONS would have catastrophic consequences */
```
Comment (patacongo) [inline mm/mm_heap/mm_initialize.c]: Blank line required after all comments.

**Comment by davids5** [comment]
Greg we have him run the tools on this in its correct form.

**Comment by patacongo** [comment]
@davids5 This change MUST BE DECLINED.  It makes extensive modification too tools/Makefile.unix, tools/Makefile.win, and tools/Config.mk.  That cannot be allowed!  The nature of the changes or also incorrect and ill-informed.  We cannot have this.

**Comment by patacongo** [comment]
This may be reopened if changes to tools/Makefile.win, Makefile.unix, and Config.mk are REMOVED.  Those are shit changes and have nothing to do with STM32H7 FMC.

**Comment by davids5** [comment]
ehmm @patacongo - Do you want more contributions? Please read what I wrote.  

>Thank you. These all have great merit. But the coupling is problematic. In looking at this it should be 3 PRs for the 3 commits.

File: arch/arm/src/stm32h7/stm32_fmc.c (hunk 1 above)
Comment (anpaza) [reply arch/arm/src/stm32h7/stm32_fmc.c]: Sorry, its a leftover from the merge from my other git repo, and forgot to remove them after checking that it works. Will fix that now.

**Pull request closed without merging**